### PR TITLE
[TP] Enable more generic attn in Tensor Parallelism

### PR DESCRIPTION
**Pull request by fduwjj**
To make TP more generic for Attention module, we come up with this new col/rowwise parallel style.

Basically, the idea behind is that:
We only do DTensor op for Col/Rowwise sharded part. For the rest of ATen ops, we will leave it to Tensor ops.

And we set this behavior as default for Colwise and Rowwise parallel style. If people want to customize it, they can always pass in different prepare_input or prepare_output

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100508
* #100137

### CODE REVIEW

File: torch/distributed/tensor/parallel/style.py [hunk 1]
```diff
@@ -388,3 +325,67 @@ def make_output_reshard_tensor(
...
+    Partitioning the row of a module.
+    We assume the input to be a sharded :class:`DTensor` and output to be a replicated :class:`DTensor`.
+    We don't perform reshard for input.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(None, make_output_replicate_1d)
+
+
+class ColwiseParallelForAttn(ColwiseParallel):
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:372]: Can we just merge this to `ColwiseParallel` itself? Maybe add a init arg if you want to distinguish difference of input/output control

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -388,3 +325,67 @@ def make_output_reshard_tensor(
...
+    Partitioning the column of a tensor or module.
+    We assume the input to be a replicated :class:`DTensor` and output to be a sharded :class:`DTensor`.
+    We don't perform reshard for output.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(make_input_replicate_1d, make_sharded_output_tensor)
+
+
+class RowwiseParallelForAttn(RowwiseParallel):
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:383]: Let's try to make sure we only have `RowwiseParallel` and `ColwiseParallel`, and `PairwiseParallel`, I feel adding a new style with subtle difference on input/output might not be a good UX.

**Review by wanchaol — COMMENTED** [review]
First pass on this PR, mainly I would like to see if we can consolidate those different styles. imo exposing different style variants could really confuse the user if the style is not quite distinguishable from others.

File: torch/distributed/tensor/parallel/style.py [hunk 3]
```diff
@@ -66,7 +65,7 @@ def __init__(self, _prepare_input=None, _prepare_output=None) -> None:
         super().__init__(_prepare_input, _prepare_output)
 
 
-class PairwiseSequenceParallel(PairwiseParallel):
+class SequenceParallel(PairwiseParallel):
     """
     PairwiseSequenceParallel concatenate colwise and rowwise styles as a fixed
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:70]: can we also fix the docs so to the correct name?

File: torch/distributed/tensor/parallel/style.py [hunk 4]
```diff
@@ -320,6 +277,28 @@ def make_output_tensor(
     ).to_local()  # type: ignore[call-arg]
 
 
+@_prepare_output_validate  # type: ignore[arg-type] # pyre-ignore[56]
+def make_sharded_output_tensor(
+    output: DTensor, _device_mesh: Optional[DeviceMesh] = None
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:282]: is there a reason we pass in `_device_mesh`? I thought we don't need this to convert a DTensor to torch.Tensor?

File: torch/distributed/tensor/parallel/style.py [hunk 5]
```diff
@@ -320,6 +277,28 @@ def make_output_tensor(
...
+) -> torch.Tensor:
+    """
+    Convert sharded Output DTensor to Tensor.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:285]: nit: to `torch.Tensor`

File: torch/distributed/tensor/parallel/style.py [hunk 6]
```diff
@@ -320,6 +277,28 @@ def make_output_tensor(
...
+
+    Args:
+        output (:class:`DTensor`):
+            Output of module to be converted.
+        _device_mesh (:class:`DeviceMesh`, optional):
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:290]: Oh I see, maybe we don't really need to document this argument? Added a comment to explain this should be enough

File: torch/distributed/tensor/parallel/style.py [hunk 7]
```diff
@@ -342,3 +321,23 @@ def make_output_reshard_tensor(
     """
 
     return make_output_shard_1d(output, device_mesh).to_local()  # type: ignore[call-arg, attr-defined]
+
+
+class RowwiseParallel(ParallelStyle):
+    """
+    Partitioning the row of a module.
+    We assume the input to be a sharded :class:`DTensor` and output to be a replicated :class:`DTensor`.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:329]: nit: we should update the doc to reflect the fact it's returning a torch.Tensor

File: torch/distributed/tensor/parallel/style.py [hunk 8]
```diff
@@ -342,3 +321,23 @@ def make_output_reshard_tensor(
...
+
+    def __init__(self, _prepare_input=make_input_shard_1d_last_dim, _prepare_output=make_output_tensor) -> None:
+        super().__init__(_prepare_input, _prepare_output)
+
+
+class ColwiseParallel(ParallelStyle):
+    """
+    Partitioning the column of a tensor or module.
+    We assume the input to be a replicated :class:`DTensor` and output to be a sharded :class:`DTensor`.
+    """
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:340]: ditto: sharded torch.Tensor

**Review by wanchaol — APPROVED** [review]
lgtm, we should update the docs to reflect the changes before landing

**Review by wanchaol — APPROVED** [review]
lgtm, we should update the docs to reflect the changes before landing

File: torch/distributed/tensor/parallel/style.py (hunk 3 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:70]: oh sure.

File: torch/distributed/tensor/parallel/style.py (hunk 4 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:282]: right, but we do explicitly pass in here:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L531

**Pull request closed without merging**